### PR TITLE
use FollowOn for CWLJobWrapper for correct scheduling

### LINKED ISSUE #2048: cwl dynamic resource requirement bug
I am trying use the dynamic resource requirement feature in gridEngine which was merged in #1810 and found an issue. 
To reproduce the issue, please add the below lines to toil/src/toil/test/cwl/revtool.cwl  and run "cwltoil --no-container --retryCount 0  --batchSystem gridEngine revsort.cwl revsort-job.json"
``` cwl
requirements:
  - class: InlineJavascriptRequirement
  - class: ResourceRequirement
    ramMin: $(inputs.input.size)
```

Below is the worker log for sorttool

```python
---TOIL WORKER OUTPUT LOG---
INFO:toil:Running Toil version 3.13.0-3cb535425306955028ed7604114f772990d00440.
Traceback (most recent call last):
  File "/work/toil/src/toil/worker.py", line 301, in main
    job = Job._loadJob(jobGraph.command, jobStore)
  File "/work/toil/src/toil/job.py", line 902, in _loadJob
    return cls._unpickle(userModule, fileHandle, jobStore.config)
  File "/work/toil/src/toil/job.py", line 931, in _unpickle
    runnable = unpickler.load()
  File "/work/toil/src/toil/job.py", line 1735, in __new__
    return cls._resolve(*args)
  File "/work/toil/src/toil/job.py", line 1747, in _resolve
    value = pickle.load(fileHandle)
  File "/work/toil/src/toil/job.py", line 1735, in __new__
    return cls._resolve(*args)
  File "/work/toil/src/toil/job.py", line 1747, in _resolve
    value = pickle.load(fileHandle)
  File "/work/toil/src/toil/job.py", line 1827, in __setstate__
    "this job as a child/follow-on of {jobName}.".format(jobName=jobName))
RuntimeError: (RuntimeError(RuntimeError("This job was passed a promise that wasn't yet resolved when it ran. The job 'file:///work/toil/src/toil/test/cwl/revtool.cwl' rev that fulfills this promise hasn't yet finished. This means that there aren't enough constraints to ensure the current job always runs after 'file:///work/toil/src/toil/test/cwl/revtool.cwl' rev. Consider adding a follow-on indirection between this job and its parent, or adding this job as a child/follow-on of 'file:///work/toil/src/toil/test/cwl/revtool.cwl' rev.",), <class 'toil.job.Promise'>, (u'file:/tmp/tmp14goWU', 'N/u/tmpF1_VvA.tmp')), <class 'toil.job.Promise'>, (u'file:/tmp/tmp14goWU', 'O/g/tmp_Mr8W3.tmp'))
ERROR:toil.worker:Exiting the worker because of a failed job on host host01
WARNING:toil.jobGraph:Due to failure we are reducing the remaining retry count of job 'file:///work/toil/src/toil/test/cwl/sorttool.cwl' sort J/b/jobhOqgt1 with ID J/b/jobhOqgt1 to 0

```

As per my understanding, there are two steps revtool and sorttool in the revsort workflow. Two wrapper jobs are created for both of them. The revtool_wrapper spawns the actual revtool job, but also sorttool_wrapper job is started. Since sorttool wrapper does not have its promise fulfilled the workflow fails.

### PR DESCRIPTION
**Pull request by mr-c**
Fix #2048 , now with a test to prove it!

### CODE REVIEW

**Comment by mr-c** [comment]
@pkp124 Does this work for you with gridEngine?

**Comment by pkp124** [comment]
@mr-c , this works for me! Thanks!

File: src/toil/cwl/cwltoil.py [hunk 1]
```diff
@@ -406,18 +403,18 @@ class CWLJob(Job):
...
 
     def __init__(self, tool, cwljob, **kwargs):
+        self.cwltool = remove_pickle_problems(tool)
         if 'builder' in kwargs:
             builder = kwargs["builder"]
         else:
             builder = cwltool.builder.Builder()
             builder.job = cwljob
-            builder.requirements = []
+            builder.requirements = self.cwltool.requirements
```
Comment (mr-c) [inline src/toil/cwl/cwltoil.py]: Needed to grab `InlineJavascriptRequirement` or other relevant requirements

**Comment by mr-c** [comment]
Confirmed for LSF as well :tada:

**Comment by mr-c** [comment]
And I'm adding a new CWL conformance test for this as well: https://github.com/common-workflow-language/common-workflow-language/pull/626/

**Comment by pkp124** [comment]
I suppose this is true for every batchsystem, including singlemachine

**Comment by mr-c** [comment]
@pkp124 You'd be surprised :-) BTW, could you do me a favor: run the CWL conformance tests using `toil-cwl-runner` and gridengine? I did it using Toil's Jenkins instance with their emulation setup, but I would feel better if it was run at least once on a real cluster

Something like https://github.com/EBI-Metagenomics/ebi-metagenomics-cwl/issues/57#issuecomment-304446975 
You may benefit from https://github.com/indigo-dc/udocker and a local nodejs installation

**Comment by pkp124** [comment]
Ran the conformance tests, with the change. As feared, test 36 and 39 fail, which are related to the scatter requirement, with the problem of promise not fulfilled. The cases are:
- v1.0/scatter-wf1.cwl v1.0/scatter-empty-job1.json
- 'v1.0/scatter-wf3.cwl#main' v1.0/scatter-empty-job3.json

**Comment by mr-c** [comment]
@pkp124 Can you email me your logs? The conformance tests with this branch pass for me on LSF.

I'm mrc@commonwl.org.

**Comment by pkp124** [comment]
So, a second run of conformance test on a clean "cwljobwrapper_followon" branch. Following are the failed cases:
1. v1.0/count-lines3-wf.cwl v1.0/count-lines3-job.json 
2. v1.0/scatter-wf2.cwl v1.0/scatter-empty-job2.json
3. v1.0/scatter-wf1.cwl v1.0/scatter-empty-job1.json
4.  'v1.0/scatter-wf3.cwl#main' v1.0/scatter-empty-job3.json

In all cases, the job 'CWLGather' that fulfills this promise hasn't yet finished.

**Comment by pkp124** [comment]
@mr-c , i have mailed you the log

**Comment by DailyDreaming** [comment]
@mr-c What's the status on this PR?

**Comment by mr-c** [comment]
@DailyDreamin This branch works for me (and the conformance tests mentioned pass) using a grid engine cluster whose `qstat -help` reports `OGS/GE 2011.11p1`.

@pkp124 What version of grid engine are you running?

**Comment by pkp124** [comment]
@mr-c I am using SGE 8.1.8

**Comment by mr-c** [comment]
@pkp124 Thanks. So we're testing with two different grid engine codebases, which makes this tricky 😋
(I have access to Open Grid Scheduler/Grid Engine)

**Comment by mr-c** [comment]
@pkp124 Can you retest using Toil 3.15.0?

**Comment by pkp124** [comment]
@mr-c , tested now with two versions:
1. A clean 3.15.0 - conformance tests failed with scheduling error
2. merged relevant changes of this branch with 3.15.0 - conformance tests failed with scheduling error
more or less for the same test cases( with scatter requirement)

**Comment by pkp124** [comment]
Every scheduling error has the following message:

_RuntimeError: (RuntimeError(RuntimeError("This job was passed a promise that wasn't yet resolved when it ran. The job 'CWLGather' that fulfills this promise hasn't yet finished. This means that there aren't enough constraints to ensure the current job always runs after 'CWLGather'. Consider adding a follow-on indirection between this job and its parent, or adding this job as a child/follow-on of 'CWLGather'.",)_

**Comment by mr-c** [comment]
> merged relevant changes of this branch with 3.15.0
both 6c6191002500cee1c78af266bcc4b66ec3167691 & 2919c5bcd15e8936c76e8266972389e650cf64fc ?

**Comment by pkp124** [comment]
merged only changes in file cwltoil.py from both commits. Ran the conformance tests with the below command:
TMP=$PWD/tmp ./run_test.sh RUNNER=cwltoil EXTRA="--batchSystem gridEngine --logDebug --logFile ${PWD}/log --disableCaching --retryCount 0 --clean never --cleanWorkDir never --no-container " -j8

**Comment by mr-c** [comment]
Can you run the following single conformance test and post the full `${PWD}/log`?

```
TMP=$PWD/tmp toil-cwl-runner --batchSystem gridEngine --logDebug --logFile ${PWD}/log --disableCaching --retryCount 0 --clean never --cleanWorkDir never --no-container v1.0/v1.0/count-lines3-wf.cwl v1.0/v1.0/count-lines3-job.json
```

**Comment by pkp124** [comment]
@mr-c Sorry for the delay. I tried the suggested test case( and other test cases as well) multiple times( ~ 30 to 40 times), tried running multiple instances of the same test case in parallel, but could not reproduce the fail. However, if I run the conformance tests, atleast one testcase fails with the reported error. I have attached the generated log file for one of the failed cases. 
[log.txt](https://github.com/BD2KGenomics/toil/files/1880182/log.txt)

**Comment by pkp124** [comment]
Another log file from another test case, just to provide more information.
[log2.txt](https://github.com/BD2KGenomics/toil/files/1880234/log2.txt)

**Comment by mr-c** [comment]
Thank you @pkp124 , maybe 79f6d9b22434df93d36a445c57cc0e8de98a1179 fixes that for you?

**Comment by pkp124** [comment]
@mr-c , thanks for the support! Unfortunately the problem still persists, although the frequency reduces. When running regression tests with the above commit, the same problem appeared for a test case without scatter - "v1.0/env-wf1.cwl v1.0/env-job.json". below is the error message:

RuntimeError: (RuntimeError(RuntimeError("This job was passed a promise that wasn't yet resolved when it ran. The job 'file:///work/toil/common-workflow-language/v1.0/v1.0/env-tool1.cwl' /bin/bash -c echo $TEST_ENV that fulfills this promise hasn't yet finished. This means that there aren't enough constraints to ensure the current job always runs after 'file:///work/toil/common-workflow-language/v1.0/v1.0/env-tool1.cwl' /bin/bash -c echo $TEST_ENV. Consider adding a follow-on indirection between this job and its parent, or adding this job as a child/follow-on of 'file:///work/toil/common-workflow-language/v1.0/v1.0/env-tool1.cwl' /bin/bash -c echo $TEST_ENV.",), <class 'toil.job.Promise'>, (u'file:/work/toil/common-workflow-language/tmp5/tmpgu_GrH', 'P/7/tmpZB7vFJ.tmp')), <class 'toil.job.Promise'>, (u'file:/work/toil/common-workflow-language/tmp5/tmpgu_GrH', u'K/J/tmpxH5mpC.tmp'))

**Comment by pkp124** [comment]
The problem seems more fundamental to me. There might be a synchronization issue between single machine job queue(which is used to handle CWL_INTERNAL_JOBS) and grid engine job queue. I ran some tests by disabling CWL_INTERNAL_JOBS and the problem does not seem to appear any more (after running conformance tests for about 5 times).

**Comment by mr-c** [comment]
@pkp124 that's a great insight, I hadn't thought of that being an issue.

**Comment by mr-c** [comment]
@pkp124 You are right, cwltool also has that error: https://github.com/common-workflow-language/cwltool/issues/760

**Pull request merged**